### PR TITLE
[gz-cmake3] Fix export of static deps

### LINKED ISSUE #34589: [gz-common5] missing find_dependency
**Describe the bug**
missing find_dependency to gz-common5 as needed by that port's exported CMake config. (Once for all consumers.)

**Environment**
- OS: Windows / Linux
- Compiler: gcc / msvc

**To Reproduce**
1. ./vcpkg install gz-common5
2. consume gz-common5 in other port: `find_package(gz-common5)`

**Expected behavior**
It needed to find all the dependencies

**Failure logs**
-(please attached failure logs)
Missing the `GTS::GTS` and `GDAL::GDAL`

```
CMake Warning at /home/tal/vcpkg/installed/x64-linux/share/minizip/minizipConfig.cmake:1 (message):
  'find_package(minizip CONFIG)' is deprecated.  Please use
  'find_package(unofficial-minizip CONFIG)' instead.
Call Stack (most recent call first):
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/assimp/assimpConfig.cmake:32 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/FindGzAssimp.cmake:23 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-graphics/gz-common5-graphics-config.cmake:106 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-geospatial/gz-common5-geospatial-config.cmake:108 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-all/gz-common5-all-config.cmake:61 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/GzFindPackage.cmake:243 (find_package)
  CMakeLists.txt:40 (gz_find_package)


CMake Deprecation Warning at /home/tal/vcpkg/installed/x64-linux/share/rapidjson/RapidJSONConfig.cmake:3 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
Call Stack (most recent call first):
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/assimp/assimpConfig.cmake:36 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/FindGzAssimp.cmake:23 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-graphics/gz-common5-graphics-config.cmake:106 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-geospatial/gz-common5-geospatial-config.cmake:108 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-all/gz-common5-all-config.cmake:61 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/GzFindPackage.cmake:243 (find_package)
  CMakeLists.txt:40 (gz_find_package)


CMake Warning (dev) at /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/GzFindPackage.cmake:456 (message):
   -- THIS MESSAGE IS INTENDED FOR GZ-PHYSICS AUTHORS --
      (IF YOU SEE THIS, PLEASE REPORT IT)

  Could not find pkg-config information for DART.  It was not provided by the
  find-module for the package, nor was it explicitly passed into the call to
  gz_find_package(~).  This is most likely an error in this project's use of
  gz-cmake.
Call Stack (most recent call first):
  CMakeLists.txt:71 (gz_find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/GzFindPackage.cmake:456 (message):
   -- THIS MESSAGE IS INTENDED FOR GZ-PHYSICS AUTHORS --
      (IF YOU SEE THIS, PLEASE REPORT IT)

  Could not find pkg-config information for GzBullet.  It was not provided by
  the find-module for the package, nor was it explicitly passed into the call
  to gz_find_package(~).  This is most likely an error in this project's use
  of gz-cmake.
Call Stack (most recent call first):
  CMakeLists.txt:84 (gz_find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "SWSCALE::SWSCALE"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVDEVICE::AVDEVICE"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVFORMAT::AVFORMAT"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVCODEC::AVCODEC"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVUTIL::AVUTIL"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "SWSCALE::SWSCALE"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVDEVICE::AVDEVICE"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVFORMAT::AVFORMAT"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVCODEC::AVCODEC"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0111 is not set: An imported target missing its location property
  fails during generation.  Run "cmake --help-policy CMP0111" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  IMPORTED_LOCATION not set for imported target "AVUTIL::AVUTIL"
  configuration "Debug".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /home/tal/vcpkg/installed/x64-linux/share/gz-common5-graphics/gz-common5-graphics-targets.cmake:60 (set_target_properties):
  The link interface of target "gz-common5::gz-common5-graphics" contains:

    GTS::GTS

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-graphics/gz-common5-graphics-config.cmake:113 (include)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-geospatial/gz-common5-geospatial-config.cmake:108 (find_package)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-all/gz-common5-all-config.cmake:61 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/GzFindPackage.cmake:243 (find_package)
  CMakeLists.txt:40 (gz_find_package)


CMake Error at /home/tal/vcpkg/installed/x64-linux/share/gz-common5-geospatial/gz-common5-geospatial-targets.cmake:60 (set_target_properties):
  The link interface of target "gz-common5::gz-common5-geospatial" contains:

    GDAL::GDAL

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-geospatial/gz-common5-geospatial-config.cmake:112 (include)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5-all/gz-common5-all-config.cmake:61 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /snap/cmake/1336/share/cmake-3.27/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
  /home/tal/vcpkg/installed/x64-linux/share/gz-common5/gz-common5-config.cmake:193 (find_dependency)
  /home/tal/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  /home/tal/vcpkg/installed/x64-linux/share/cmake/gz-cmake3/cmake3/GzFindPackage.cmake:243 (find_package)
  CMakeLists.txt:40 (gz_find_package)


CMake Warning:
  Manually-specified variables were not used by the project:

    FETCHCONTENT_FULLY_DISCONNECTED
    VCPKG_PLATFORM_TOOLSET
    VCPKG_SET_CHARSET_FLAG
    _VCPKG_ROOT_DIR


CMake Generate step failed.  Build files cannot be regenerated correctly.
```

**Additional context**
From @dg0yt comment
https://github.com/microsoft/vcpkg/pull/33919#discussion_r1361613928

### PR DESCRIPTION
**Pull request by dg0yt**
Fixes https://github.com/microsoft/vcpkg/issues/34589.

The gz port comes with their own set of cmake functions for dependency import/export.
This PR adds fixes and hooks early in the dependency chain so that downstream usage is easier.

### CODE REVIEW

File: ports/gz-cmake3/dependencies.patch [hunk 1]
```diff
@@ -20,3 +20,16 @@ diff --git a/cmake/FindGzBullet.cmake b/cmake/FindGzBullet.cmake
...
+diff --git a/cmake/GzFindPackage.cmake b/cmake/GzFindPackage.cmake
+index 4eb7a0d..231725b 100644
+--- a/cmake/GzFindPackage.cmake
++++ b/cmake/GzFindPackage.cmake
+@@ -388,7 +388,7 @@ macro(gz_find_package PACKAGE_NAME_)
+       # ${component}_CMAKE_DEPENDENCIES variables that are specific to those
+       # componenets
+       foreach(component ${gz_find_package_REQUIRED_BY})
+-        if(NOT ${component}_${PACKAGE_NAME}_PRIVATE)
++        if(NOT ${component}_${PACKAGE_NAME}_PRIVATE OR NOT BUILD_SHARED_LIBS)
```
Comment (dg0yt) [inline ports/gz-cmake3/dependencies.patch:32]: Or simply `if(1)`. Project-wide dependencies are always exported. Only per-component dependencies are guarded like this, and the original guard ignores resolving targets for static link libs.

File: ports/gz-cmake3/dependencies.patch [hunk 2]
```diff
@@ -1,22 +1,110 @@
...
++  include("${CMAKE_CURRENT_LIST_DIR}/vcpkg/${PACKAGE_NAME}.cmake" OPTIONAL)
+   #------------------------------------
+   # Construct the arguments to pass to find_package
+   set(${PACKAGE_NAME}_find_package_args ${PACKAGE_NAME})
+@@ -192,6 +193,8 @@ macro(gz_find_package PACKAGE_NAME_)
+ 
+   if(gz_find_package_EXACT)
+     list(APPEND ${PACKAGE_NAME}_find_package_args EXACT)
++  elseif(gz_find_package_VERSION AND NOT gz_find_package_PKGCONFIG_VER_COMPARISON)
++    set(gz_find_package_PKGCONFIG_VER_COMPARISON >=)
```
Comment (dg0yt) [inline ports/gz-cmake3/dependencies.patch:98]: This just aligns cmake and and pkg-config behaviour (if pkg-config behaviour is not given explicitly). Needed when vcpkg adds the `PKGCONFIG` parameter, e.g. for GDAL.

**Comment by dg0yt** [comment]
@talregev Please test this. For gz-common5-graphics-config.cmake  x64-linux, this exports now:
~~~cmake
if(NOT gz-common5-graphics_FIND_QUIETLY)
  message(STATUS "Searching for dependencies of gz-common5-graphics")
endif()
find_package(gz-math7 ${gz_package_quiet} ${gz_package_required})
find_package(TINYXML2 ${gz_package_quiet} ${gz_package_required})
find_package(FreeImage 3.9 ${gz_package_quiet} ${gz_package_required})
find_package(GTS ${gz_package_quiet} ${gz_package_required})
find_package(GzAssimp ${gz_package_quiet} ${gz_package_required})
if(NOT gz-common5_CONFIG_INCLUDED)
  find_package(gz-common5 5.4.1 EXACT ${gz_package_quiet} ${gz_package_required})
endif()
~~~
Note that this relies on gz-cmake3's Find modules, and in the case of GTS this needs pkg-config (or pkgconf).

**Comment by talregev** [comment]
I will test it with gz-physics6 and gz-rendering7
Maybe we should add pkgconf depending on gz-common5

**Comment by talregev** [comment]
@dg0yt Thank you for your hard work. It working on both ports.

**Comment by dg0yt** [comment]
> Maybe we should add pkgconf depends for gz-common5.

This was on my todo list. Done.

**Review by LilyWangLL — APPROVED** [review]

**Pull request merged**